app/src/main/java/com/github/se/cyrcle/model/network/NetworkReceiver.kt

```diff
@@ -0,0 +1,48 @@
+package com.github.se.cyrcle.model.network
+
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.Network
+import android.net.NetworkCapabilities
+import android.net.NetworkRequest
+import android.util.Log
+import kotlinx.coroutines.flow.MutableStateFlow
+
+class NetworkReceiver(context: Context) {
+  // MutableStateFlow to keep track of the network connection status
+  private val _isConnected = MutableStateFlow(false)
+  val isConnected: MutableStateFlow<Boolean> = _isConnected
+
+  private val connectivityManager =
+      context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+  private val networkCallback =
+      object : ConnectivityManager.NetworkCallback() {
+        override fun onAvailable(network: Network) {
+          Log.d("NetworkReceiver", "Network isConnected: ${isConnected.value}")
+          _isConnected.value = true
+        }
+
+        override fun onLost(network: Network) {
+          Log.d("NetworkReceiver", "Network lost")
+          // Set isConnected to false when the network is lost
+          _isConnected.value = false
+        }
+
+        override fun onCapabilitiesChanged(
+            network: Network,
+            networkCapabilities: NetworkCapabilities
+        ) {
+          val isConnected =
+              networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+          Log.d("NetworkReceiver", "Network capabilities changed, isConnected: $isConnected")
+          _isConnected.value = isConnected
+        }
+      }
+
+  fun register() {
+    val networkRequest =
+        NetworkRequest.Builder().addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET).build()
+    connectivityManager.registerNetworkCallback(networkRequest, networkCallback)
+  }
+}
```

Comment:
Shouldn't we also have the following `fun unregister() {
      connectivityManager.unregisterNetworkCallback(networkCallback)
  }  ` ? Im not sure but it could prevent unnecessary system resource uses and maybe crashs that the callbacks could cause.

Reply:
I tried having an unregister() but it was useless since we never want to unregister the listener.
 It will automatically stop when we close the app

Reply:
I think when you kill the app android does it automatically actually. But if you dont kill the app, but let's say just go on another app while ours is still in the background, maybe you should unregister()...

Reply:
Where should I call it then ?

Reply:
Drop the case 🔫

Reply:
2 bucks

Reply:
Damn even cheaper than your mom, nice

Reply:
fuck you ni-

Reply:
Nevermind, as long as there's no bugs, let's not bother. let's just keep that in mind in case we see weird behaviours sometime in the future.